src/crimson/osd/heartbeat.h

```diff
@@ -166,33 +172,55 @@ class Heartbeat::Peer {
     }
   }
 
-  void connect();
-  void disconnect();
+  // a session starts when con_front and con_back are both connected
+  void start_session();
+  // a session resets when either con_front or con_back is reset
+  void reset_session();
+  // notify when con_front becomes ready, possibly start session
+  void notify_front_ready();
+  void connect_front();
+  // notify when con_back becomes ready, possibly start session
+  void notify_back_ready();
+  void connect_back();
+
+  void do_send_heartbeat(clock::time_point now,
+                         ceph::signedspan mnow,
+                         std::vector<seastar::future<>>*);
 
  private:
   Heartbeat& heartbeat;
   const osd_id_t peer;
 
-  /// peer connection (front)
-  crimson::net::ConnectionRef con_front;
-  /// peer connection (back)
-  crimson::net::ConnectionRef con_back;
-  /// time we sent our first ping request
+  // time we sent our first ping request
   clock::time_point first_tx;
-  /// last time we sent a ping request
+  // last time we sent a ping request
   clock::time_point last_tx;
-  /// last time we got a ping reply on the front side
+  // last time we got a ping reply on the front side
   clock::time_point last_rx_front;
-  /// last time we got a ping reply on the back side
+  // last time we got a ping reply on the back side
   clock::time_point last_rx_back;
-  /// most recent epoch we wanted this peer
+  // most recent epoch we wanted this peer
   epoch_t epoch;
 
+  // if racing happened
+  bool has_racing = false;
+  // peer connection (front)
+  crimson::net::ConnectionRef con_front;
+  bool front_ready = false;
+  // peer connection (back)
+  crimson::net::ConnectionRef con_back;
+  bool back_ready = false;
+
+  // start to send pings and track ping_history
+  bool session_started = false;
```

Comment:
i'd suggest replace front_ready, back_ready with an `unsigned connected`, and increment this number either when a front or back connection to peer is established, and decrement it when a connection is reset.

Reply:
Not sure, but seems to me the value `unsigned connected = 1` cannot represent whether the front or the back is ready. For example in `Heartbeat::Peer::handle_reset()` L527 L557, there are requirements to differentiate them.